src/status_im2/contexts/add_new_contact/events.cljs

```diff
@@ -6,114 +6,133 @@
+;; - pure functions that transform maps
+;; 3. Tests
+;; - focus on the core decision-making functions first (validate-contact)
+;; - followed by the (hopefully mostly empty) event handlers
+;; - followed by int test (add-contact-test) for status-go & downstream events
+;; 4. Testing event handlers/callbacks
+;; - remove all logic from callback: callback only dispatches
+;; - tests replace the callback with a fn that just returns args
+
+(defn init-contact ; in app-db as [:contacts/new-identity]
+  ([]
+   {:user-public-key nil ; user's public key (not the contact)
+    :input           nil ; raw user input (untrimmed)
+    :id              nil ; public-key|compressed-key|ens
+    :type            nil ; :empty|:public-key|:compressed-key|:ens
+    :ens             nil ; id.eth|id.ens-stateofus
+    :public-key      nil ; public-key (from decompression or ens resolution)
+    :state           nil ; :empty|:invalid|:decompress-key|:resolve-ens|:valid
+    :msg             nil ; keyword i18n msg
+   })
+  ([kv] (-> (init-contact) (merge kv))))
+
+(defn ->id
+  [{:keys [input] :as contact}]
+  (let [regex         #"^https?://join.status.im/u/(.+)"
+        trimmed-input (utils/safe-trim input)]
+    (->> {:id (if (empty? trimmed-input)
+                nil
+                (if-some [[_ id] (re-matches regex trimmed-input)]
+                  id
+                  trimmed-input))}
+         (merge contact))))
+
+(defn ->type
+  [{:keys [id] :as contact}]
+  (->> (cond
+         (empty? id)                  {:type :empty}
+         (v/valid-public-key? id)     {:type       :public-key
+                                       :public-key id}
+         (v/valid-compressed-key? id) {:type :compressed-key}
+         :else                        {:type :ens
+                                       :ens  (stateofus/ens-name-parse id)})
+       (merge contact)))
+
+(defn ->state
+  [{:keys [id type public-key user-public-key] :as contact}]
+  (->> (cond
+         (empty? id)                            {:state :empty}
+         (= type :public-key)                   {:state :invalid
+                                                 :msg   :t/not-a-chatkey}
+         (= public-key user-public-key)         {:state :invalid
+                                                 :msg   :t/can-not-add-yourself}
+         (and (= type :compressed-key)
+              (empty? public-key))              {:state :decompress-key}
+         (and (= type :ens)
+              (empty? public-key))              {:state :resolve-ens}
+         (and (or (= type :compressed-key) (= type :ens))
+              (v/valid-public-key? public-key)) {:state :valid})
+       (merge contact)))
+
+(def validate-contact (comp ->state ->type ->id))
+
+(defn cb [event contact] (fn [arg] (rf/dispatch [event contact arg])))
+
+(rf/defn set-new-identity
+  {:events [:contacts/set-new-identity]}
+  [{:keys [db]} input]
+  (let [upk             (get-in db [:multiaccount :public-key])
```

Comment:
what is upk ?

Reply:
Thanks, yes, i forgot to update that. I was thinking of it as 'user-public-key', which I'll replace it with.

Reply:
done